cpp/src/arrow/io/file.cc

```diff
@@ -167,8 +169,23 @@ class OSFile {
     if (length < 0) {
       return Status::IOError("Length must be non-negative");
     }
+#ifdef __linux__
+    if (reuse_) {
+      return ::arrow::internal::FileWrite(fd_.fd(),
+                                          reinterpret_cast<const uint8_t*>(data), length);
```

Comment:
This line can probably be hoisted above #ifdef, leave only one copy, not three as in current code.